### PR TITLE
Addon-docs: Refactor DocsRender/Context

### PR DESCRIPTION
**Pull request by tmeasday**
Telescoping on https://github.com/storybookjs/storybook/pull/18634

## What I did

Refactored `DocsContextProps` to be a little bit simpler, and much more consistently implemented across the three different docs contexts:

 - **Standalone** - `.mdx` files that may have a primary CSF file (`<Meta of={}>`) as well as reference others (`<Story ... meta={}>`)
 - **Template** - `*.stories.mdx` files or docs page entries that are attached to one (or more, in the case of component stories spread across CSF files) CSF files. (NOTE this is the legacy case)
 - **External** - similar to Standalone, but running in an external system.

The APIs available to doc blocks are backwards compatible, with the following changes:
 - `storyById` can now be called with no parameters to get the primary story. This is preferred (and the blocks are changed) as the `context.id` may not be a story id in the standalone or external case.
 - `mdxStoryNameToKey/mdxComponentAnnotations` are no longer defined. There is a new API available, `context.storyIdByName` that serves the same purpose. It is used by our blocks to get ids for MDX defined stories. **This is a breaking change**, however I am not sure if it was ever used externally.

Now that the MDX compilers don't need to append `mdxStoryNameToKey/mdxComponentAnnotations` any more, there's no need for the `AddContext` utility either. I've got a task to do this cleanup later: https://linear.app/chromaui/issue/SB-465/remove-call-to-addtemplate-from-mdx-generators

### CODE REVIEW

File: lib/preview-web/src/PreviewWeb.tsx [hunk 1]
```diff
@@ -389,33 +388,6 @@ export class PreviewWeb<TFramework extends AnyFramework> extends Preview<TFramew
...
-      story
-    );
-    render.renderToElement(element);
-
-    this.storyRenders.push(render);
-
-    return async () => {
-      await this.teardownRender(render);
-    };
-  }
```
Comment (tmeasday) [inline lib/preview-web/src/PreviewWeb.tsx:417]: This was duplicated from the super class, I think from a merge conflict.

**Comment by tmeasday** [comment]
@ndelangen I got a flaky test run: https://app.circleci.com/pipelines/github/storybookjs/storybook/27512/workflows/ea52cb29-f334-4cb2-9192-f7c8df1c1bf6/jobs/394112

I tracked down the issue -- the problem is the project `react_legacy_root_api`, which sets (in `main.js`):

```js
reactOptions: {
  legacyRootApi: true,
},
```

No longer gets `FRAMEWORK_OPTIONS` set in the preview, which means it renders in react 18 mode, which makes this test flaky. I'll discuss the flakiness below. But @ndelangen I assume the lack of globals in the preview is due to changes in setup that you might know about?

---

Regarding the flakiness. The reason is that the new `react-dom/client` [API no longer has a "done" callback](https://github.com/reactwg/react-18/discussions/5), so now we have a simple `setTimeout(,0)` before assuming the story is rendered:

https://github.com/storybookjs/storybook/blob/849cfcaed2bade811b429e46893bbb50d2da84a1/renderers/react/src/render.tsx#L42-L44

@valentinpalkovic I think we are going to need to come up with a better solution than that. The problem is that our "storybook hooks" that e.g. [run inside the JSX decorator](https://github.com/storybookjs/storybook/blob/849cfcaed2bade811b429e46893bbb50d2da84a1/renderers/react/src/docs/jsxDecorator.tsx#L185-L189) rely on the current story still being in the "render" phase at the time the hook is called to work properly.

**Comment by ndelangen** [comment]
@tmeasday I don't see why that config/setting would be missing. That code didn't change AFAIK, the code is here:
https://github.com/storybookjs/storybook/blob/5683419d4b6953c9bb3624e00be6bc00437c4fae/lib/builder-webpack5/src/preview/iframe-webpack.config.ts#L206-L218

**Comment by ndelangen** [comment]
So that might mean getting the value coming from preset changed somehow.. I'm investigating.

**Comment by ndelangen** [comment]
I figured it out... opening PR soon

**Review by shilman — APPROVED** [review]
Nice one 🙌

File: lib/blocks/src/blocks/external/ExternalDocsContainer.tsx [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+import { ThemeProvider, themes, ensure } from '@storybook/theming';
+import { AnyFramework } from '@storybook/csf';
+
+import { DocsContext } from '../DocsContext';
+import { ExternalPreview } from './ExternalPreview';
+
+let preview: ExternalPreview<AnyFramework>;
+
+export const ExternalDocsContainer: React.FC<{ projectAnnotations: any }> = ({
```
[the remark this answers is not in the document]
Comment (tmeasday) [reply lib/blocks/src/blocks/external/ExternalDocsContainer.tsx:11]: I'll sort this out in an upcoming PR.

File: lib/blocks/src/blocks/external/ExternalDocsContainer.tsx [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  projectAnnotations,
+  children,
+}) => {
+  if (!preview) preview = new ExternalPreview(projectAnnotations);
+
+  return (
+    <DocsContext.Provider value={preview.docsContext()}>
+      <ThemeProvider theme={ensure(themes.light)}>{children}</ThemeProvider>
```
[the remark this answers is not in the document]
Comment (tmeasday) [reply lib/blocks/src/blocks/external/ExternalDocsContainer.tsx:19]: Separate PR. https://linear.app/chromaui/issue/SB-453/dependencies-ensure-react-dom-mdx-dependencies-come-from-addon-docs

**Pull request merged**